### PR TITLE
Move save_checkpoint() out of train.py and refactor checkpoint saving…

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
… in all trainers

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Training logs after this change:
```
## steps=10 profiler=xplane checkpoint_period=7

completed step: 0, seconds: 28.332, TFLOP/s/device: 1.491, Tokens/s/device: 36.142, total_weights: 3573, loss: 0.993
completed step: 1, seconds: 1.099, TFLOP/s/device: 38.454, Tokens/s/device: 932.156, total_weights: 3253, loss: 0.993
completed step: 2, seconds: 0.999, TFLOP/s/device: 42.289, Tokens/s/device: 1025.108, total_weights: 1692, loss: 0.912
completed step: 3, seconds: 0.994, TFLOP/s/device: 42.493, Tokens/s/device: 1030.049, total_weights: 2944, loss: 1.105
completed step: 4, seconds: 0.993, TFLOP/s/device: 42.527, Tokens/s/device: 1030.875, total_weights: 3466, loss: 0.948
completed step: 5, seconds: 17.485, TFLOP/s/device: 2.416, Tokens/s/device: 58.564, total_weights: 2420, loss: 0.947
completed step: 6, seconds: 0.981, TFLOP/s/device: 43.052, Tokens/s/device: 1043.599, total_weights: 1655, loss: 1.059
completed step: 7, seconds: 34.116, TFLOP/s/device: 1.238, Tokens/s/device: 30.016, total_weights: 3335, loss: 1.000
completed step: 8, seconds: 0.982, TFLOP/s/device: 43.027, Tokens/s/device: 1042.997, total_weights: 1676, loss: 0.851
completed step: 9, seconds: 0.989, TFLOP/s/device: 42.720, Tokens/s/device: 1035.561, total_weights: 1676, loss: 0.965
```

Training logs before this change:
```
## steps=10 profiler=xplane checkpoint_period=7

completed step: 0, seconds: 27.544, TFLOP/s/device: 1.534, Tokens/s/device: 37.177, total_weights: 3573, loss: 0.993
completed step: 1, seconds: 1.105, TFLOP/s/device: 38.242, Tokens/s/device: 927.006, total_weights: 3253, loss: 0.993
completed step: 2, seconds: 0.996, TFLOP/s/device: 42.403, Tokens/s/device: 1027.874, total_weights: 1692, loss: 0.912
completed step: 3, seconds: 0.996, TFLOP/s/device: 42.426, Tokens/s/device: 1028.421, total_weights: 2944, loss: 1.105
completed step: 4, seconds: 0.994, TFLOP/s/device: 42.517, Tokens/s/device: 1030.628, total_weights: 3466, loss: 0.948
completed step: 5, seconds: 16.416, TFLOP/s/device: 2.573, Tokens/s/device: 62.376, total_weights: 2420, loss: 0.947
completed step: 6, seconds: 0.982, TFLOP/s/device: 43.037, Tokens/s/device: 1043.238, total_weights: 1655, loss: 1.059
completed step: 7, seconds: 35.911, TFLOP/s/device: 1.176, Tokens/s/device: 28.515, total_weights: 3335, loss: 1.000
completed step: 8, seconds: 0.981, TFLOP/s/device: 43.063, Tokens/s/device: 1043.866, total_weights: 1676, loss: 0.851
completed step: 9, seconds: 0.986, TFLOP/s/device: 42.833, Tokens/s/device: 1038.301, total_weights: 1676, loss: 0.965
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/checkpointing.py [hunk 1]
```diff
@@ -421,3 +422,90 @@ def save_params_to_path(checkpoint_dir, params, use_ocdbt=True, use_zarr3=True):
...
+  # specify chunk_byte_size to force orbax to control maximum file size in checkpoint
+  chunk_byte_size = config.checkpoint_storage_target_data_file_size_bytes if config else DEFAULT_OCDBT_TARGET_DATA_FILE_SIZE
+
+  checkpoint_args = ocp.args.PyTreeSave(
+      item=state,
+      save_args=jax.tree.map(lambda _: ocp.SaveArgs(chunk_byte_size=chunk_byte_size), state),
+      ocdbt_target_data_file_size=chunk_byte_size,
+  )
+
+  if isinstance(
```
Comment (richjames0) [inline MaxText/checkpointing.py]: How about replacing this with a match, e.g. (untested)

```
from ocp.args import Composite
from emergency_checkpoint_manager import CheckpointManager, ReplicatorCheckpointManager

match (checkpoint_manager, config):
  case (CheckpointManager() | ReplicatorCheckpointManager(), _,):
    replicator_error_handler(config)
    return checkpoint_manager.save(step, args=Composite(state=checkpoint_args),force=force)
  
  case (_, cfg) if cfg and cfg.dataset_type == "grain":
    iter_ = grain.PyGrainCheckpointSave(data_iterator.local_iterator)
    return checkpoint_manager.save(step, args=Composite(items=checkpoint_args, iter=iter_), force=force)

  case _:
    return checkpoint_manager.save(step, args=Composite(items=checkpoint_args), force=force)
    ```

File: MaxText/checkpointing.py [hunk 2]
```diff
@@ -374,9 +377,7 @@ def setup_checkpoint_logger(config) -> Any | None:  # pytype: disable=attribute-
   max_logging.log("Setting up checkpoint logger...")
   if config.enable_checkpoint_cloud_logger:
     logger_name = f"goodput_{config.run_name}"
-    options = ocp.logging.CloudLoggerOptions(
-        job_name=config.run_name, logger_name=logger_name
-    )
+    options = ocp.logging.CloudLoggerOptions(job_name=config.run_name, logger_name=logger_name)
```
Comment (richjames0) [inline MaxText/checkpointing.py]: This brings me joy

File: MaxText/checkpointing.py (hunk 1 above)
Comment (SurbhiJainUSC) [reply MaxText/checkpointing.py]: Updated, thanks!

File: MaxText/checkpointing.py [hunk 3]
```diff
@@ -164,14 +164,14 @@ def replicator_error_handler(config: Any):
       raise ValueError("Replicator fatal error found in replicator.failed file.")
 
 
-def process_replicator_error_file(error_file: str) -> bool:
+def process_replicator_error_file(error_file: str):
```
Comment (A9isha) [inline MaxText/checkpointing.py]: any reason to remove this return type?

File: MaxText/checkpointing.py [hunk 4]
```diff
@@ -354,16 +319,14 @@ def map_to_pspec(data):
...
+    restored = ocp.StandardCheckpointer().restore(p, abstract_unboxed_pre_state)
     return {"items": restored}, None
-
   else:
     max_logging.log("No existing checkpoints found, not restoring checkpoint.")
     return None, None
 
 
-def setup_checkpoint_logger(config) -> Any | None:  # pytype: disable=attribute-error
+def setup_checkpoint_logger(config):  # pytype: disable=attribute-error
```
Comment (A9isha) [inline MaxText/checkpointing.py]: similar qs

File: MaxText/checkpointing.py [hunk 5]
```diff
@@ -374,12 +337,10 @@ def setup_checkpoint_logger(config) -> Any | None:  # pytype: disable=attribute-
...
   if config.enable_checkpoint_cloud_logger:
     logger_name = f"goodput_{config.run_name}"
-    options = ocp.logging.CloudLoggerOptions(
-        job_name=config.run_name, logger_name=logger_name
+    orbax_cloud_logger = ocp.logging.CloudLogger(
+        options=ocp.logging.CloudLoggerOptions(job_name=config.run_name, logger_name=logger_name)
     )
-    orbax_cloud_logger = ocp.logging.CloudLogger(options=options)
     max_logging.log("Successfully set up checkpoint cloud logger.")
-    return orbax_cloud_logger
```
Comment (A9isha) [inline MaxText/checkpointing.py:382]: thank you for cleaning this up!

File: MaxText/checkpointing.py [hunk 6]
```diff
@@ -421,3 +384,72 @@ def save_params_to_path(checkpoint_dir, params, use_ocdbt=True, use_zarr3=True):
...
+  # If 'step' is not provided, this call is for a potential final checkpoint
+  # and use the last completed step from the state.
+  actual_step = (int(state.step) - 1) if step is None else int(step)
+
+  # Determine if a checkpoint save should be forced, overriding the usual `config.checkpoint_period` logic.
+  # This occurs if this function was called:
+  # without an explicit 'step' (implying it's a checkpoint save for final step),
+  # AND the 'actual_step' is a valid step,
+  # AND it's not a step that would normally trigger a checkpoint save.
+  force_ckpt_save = step is None and actual_step != -1 and (actual_step % config.checkpoint_period != 0)
```
Comment (A9isha) [inline MaxText/checkpointing.py:404]: say training step count = 10
checkpoint freq = 5
then we only save checkpoints at step 0 and 5 because we do % operation

do we want to incorporate the functionality that it saves the checkpoint at step 9?

File: MaxText/train.py [hunk 7]
```diff
@@ -769,28 +679,13 @@ def train_loop(config, recorder, state=None):
 
       step_time_delta = datetime.datetime.now() - step_start_time
       metric_logger.record_train_metrics(metrics, step, step_time_delta)
+
+    state_to_save = state if not config.use_dpo else _split_dpo_state(state)[0]
+    checkpointing.maybe_save_checkpoint(checkpoint_manager, state_to_save, config, data_iterator)
```
Comment (A9isha) [inline MaxText/train.py:644]: do we want to stop training if checkpointing crashes?

File: MaxText/checkpointing.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply MaxText/checkpointing.py]: Just to be consistent with other methods in the module.

File: MaxText/checkpointing.py (hunk 6 above)
Comment (SurbhiJainUSC) [reply MaxText/checkpointing.py:404]: This force_ckpt_save is for the same case you are pointing out. 
For step=9 with checkpoint freq=5, `config.checkpoint_period != 0` will be True and force_ckpt_savr=True. This condition forces checkpoint manager to save the checkpoint.

File: MaxText/checkpointing.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply MaxText/checkpointing.py]: For consistency with other methods in the module.

File: MaxText/train.py (hunk 7 above)
Comment (SurbhiJainUSC) [reply MaxText/train.py:644]: Training will stop and the program crashes if `checkpoint_manager.save()` throws any exception. This is the same functionality as we currently have in the codebase. Do you want to explicitly raise `exceptions.StopTraining` exception in that case?

File: MaxText/checkpointing.py (hunk 3 above)
Comment (bvandermoon) [reply MaxText/checkpointing.py]: Agree it is unfortunate to have the inconsistent style. I would generally prefer to keep the ones we have though. Maybe if we get time in the future (haha) we can add them in other places as well

File: MaxText/train.py [hunk 8]
```diff
@@ -723,15 +640,8 @@ def train_loop(config, recorder, state=None):
...
-        state_to_save = state if not config.use_dpo else _split_dpo_state(state)[0]
-        if save_checkpoint(checkpoint_manager, int(step), state_to_save, config.dataset_type, data_iterator, config):
-          checkpointing.print_save_message(step, config.async_checkpointing)
-
-        # Upon preemption, exit when and only when all ongoing saves are complete.
-        if checkpoint_manager.reached_preemption(step):
-          checkpoint_manager.wait_until_finished()
-          sys.exit()
+      state_to_save = state if not config.use_dpo else _split_dpo_state(state)[0]
+      checkpointing.maybe_save_checkpoint(checkpoint_manager, state_to_save, config, data_iterator, step)
```
Comment (bvandermoon) [inline MaxText/train.py:604]: Can we move the `state_to_save` logic to `maybe_save_checkpoint`? I think config has the info needed to determine which split_<>_state(...) method to call so it should also work for grpo_trainer. Thoughts?

**Review by bvandermoon — APPROVED** [review]
LGTM pending @A9isha's open comments

File: MaxText/train.py (hunk 8 above)
Comment (SurbhiJainUSC) [reply MaxText/train.py:604]: This logic is very specific to DPO and GRPO. There is a plan to cleanup DPO related code from train.py and have a separate trainer (dpo_trainer), similar to sft and grpo. So, this splitting should actually be done by dpo_trainer and grpo_trainer.

File: MaxText/train.py (hunk 7 above)
Comment (SurbhiJainUSC) [reply MaxText/train.py:644]: Added `StopTraining` exception whenever checkpointing fails.

File: MaxText/checkpointing.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply MaxText/checkpointing.py]: Reverted this change.

File: MaxText/train.py (hunk 7 above)
Comment (A9isha) [reply MaxText/train.py:644]: Thanks Surbhi!

**Review by A9isha — APPROVED** [review]
Thanks Surbhi!
